R/cube-residuals.R

```diff
@@ -23,46 +67,185 @@ standardizedMRResiduals <- function(cube) {
+    dim <- match.arg(dim)
+
+    # grab the names of extents for each dimensions
+    # TODO: we shouldn't need to do this if we can just pass baseline/x into the
+    # subsetting function when we can subset by name _or_ index natively
+    if (dim == "cols") {
+        names <- colnames(as.array(cube))
+    } else if (dim == "rows") {
+        names <- rownames(as.array(cube))
     }
 
-    if (any(startsWith(getDimTypes(model), "mr_"))) {
-        return(standardizedMRResiduals(model))
+    # convert to numeric indices
+    # TODO: remove when we can accept either characters or indices and pass them
+    # to subsetting unmolested
+    if (is.character(baseline)) {
+        baseline_ind <- which(names == baseline)
     } else {
-        array_to_test <- as.array(noTransforms(model))
-        return(chisq.test(array_to_test)$stdres)
+        stop("Currently, column comparison only accepts category names.")
     }
-})
+    if (is.character(x)) {
+        x_ind <- which(names == x)
+    } else {
+        stop("Currently, column comparison only accepts category names.")
+    }
+
+    # ensure that the extents given are in the cube
+    # TODO: remove when we can defer this check to the subsetting methods
+    not_in_cube <- !(c(baseline, x) %in% names)
+    if (any(not_in_cube)) {
+        stop(c(baseline, x)[not_in_cube], " is not a column or row in the cube")
+    }
+
+    # ensure that there are not MRs on the comparison direction
+    dim_types <- getDimTypes(cube)
+    if ((dim == "cols" & startsWith(dim_types[2], "mr_")) |
+        (dim == "rows" & startsWith(dim_types[1], "mr_"))) {
+        stop(
+            "Column or row z-scores are not implemented for multiple response ",
+            "dimensions"
+        )
+    }
+
+    # generate the 2xm or nx2 table tfor testing
+    if (dim == "cols") {
+        sub_cube <- cube[, c(x_ind, baseline_ind)]
+    } else if (dim == "rows") {
+        sub_cube <- cube[c(x_ind, baseline_ind), ]
+    }
+
+    # make the test
+    out <- zScores(sub_cube)
+
+    # only return the scores for x
+    if (dim == "cols") {
+        out <- out[, x, drop = FALSE]
+    } else if (dim == "rows") {
+        out <- out[x, , drop = FALSE]
+    }
+
+    return(out)
+}
+
+
+#' Pairwise column and row comparison
+#'
+#' Given a single baseline column compare each other row or column against this
+#' baseline. Internally this function uses `compareDims()` iteratively.
+#'
+#' *Warning* since there is more than one comparison being made against each
+#' baseline the z-scores, and especially the p-values derived from these
+#' z-scores should be interpreted with caution. Using standard p-value cutoffs
+#' will result in anti-conservative interpretations because of the multiple
+#' comparisons problem. Adjustments to p-value cut offs (e.g. Bonferonni
+#' correction) should be used when interpreting z-scores from the
+#' `compare[Rows|Cols|Dims]Pairwise()` family of functions.
```

Comment:
It might be good to include a reference (maybe a wikipedia article or the like) here. This paragraph is probably only meaningful to people who have a very good understanding of p-values, which is not that many people.